flowpm/raytracing.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Wed Feb  3 12:29:19 2021
+
+@author: Denise Lanzieri
+"""
+import numpy as np
+import tensorflow as tf
+import tensorflow_addons as tfa
+import flowpm
+import flowpm.constants as constants
+
+
+def generate_matrix():
+  """ Generate the possible rotation matrices
+    """
+  x = np.asarray([1, 0, 0], dtype=int)
+  y = np.asarray([0, 1, 0], dtype=int)
+  z = np.asarray([0, 0, 1], dtype=int)
+  M_matrices = [np.asarray([x,y,z],dtype=int), np.asarray([x,z,y],dtype=int),np.asarray([z,y,x],dtype=int),np.asarray([z,x,y],dtype=int), \
+                     np.asarray([y,x,z],dtype=int), np.asarray([y,z,x],dtype=int)]
+
+  return M_matrices
+
+
+def shift():
+  """ Generate possible xy shifts
+    """
+  I = np.zeros(3)
+  facx = np.random.uniform(0, 1)
+  facy = np.random.uniform(0, 1)
+  xyshifts = [np.asarray([facx, facy, 0], dtype=float)]
+  return I + xyshifts
+
+
+def density_plane(state,
+                  nc,
+                  center,
+                  Matrix,
+                  shift,
+                  width,
+                  plane_resolution,
+                  name='density_plane'):
+  """ Extract a slice from the input state vector and
+  project it as a density plane.
+
+  Args:
+    state: `Tensor`, input state tensor.
+    nc: int or list of int, size of simulation volume
+    center: float, center of the slice along the z axis in voxel coordinates
+    width: float, width of the slice in voxel coordinates
+    plane_size: int, number of pixels of the density plane. 
+    name: `string`, name of the operation.
+
+  Returns:
+    `Tensor` of shape [batch_size, plane_size, plane_size], of projected density plane.
+  """
+  with tf.name_scope(name):
+    state = tf.convert_to_tensor(state, name="state")
+    if isinstance(nc, int):
+      nc = [nc, nc, nc]
+    nx, ny, nz = nc
+    pos = state[0]
+
+    shape = tf.shape(pos)
+    batch_size = shape[0]
+    y = tf.einsum('ij,bkj->bki', Matrix, pos)
```

Comment:
we will probably only want to do this if `rotation` and/or `shift` is not `None`

Reply:
node